### PR TITLE
[Sources] Snowflake - Part 1

### PR DESCRIPTION
**Pull request by vellames**
This is part 1 of #5226.
A new PR will be published with more sources

### <samp>🤖 Generated by Copilot at 8922891</samp>

This pull request adds three new sources for Snowflake that emit events when a new database, schema, or table is created. It also adds a new method to the common source module for Snowflake that allows sources to run in a single process mode and perform additional processing after getting the statement. This is needed to update the last execution time for the sources.

### <samp>🤖 Generated by Copilot at 8922891</samp>

> _Snowflake sources grow_
> _`alwaysRunInSingleProcessMode`_
> _Winter databases_

### <samp>🤖 Generated by Copilot at 8922891</samp>

*  Add a new method `alwaysRunInSingleProcessMode` to the common source module for Snowflake, which determines whether the source should run in a single process mode regardless of the `emitIndividualEvents` option ([link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-8463d90b1fce2b0a62c9996d446a3b269e680bc0598a76accdadd3ea59d268d2L201-R211))
*  Add a conditional call to `this.additionalProccessing` in the common source module, which allows sources to perform custom logic after getting the statement ([link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-8463d90b1fce2b0a62c9996d446a3b269e680bc0598a76accdadd3ea59d268d2L201-R211))
*  Modify the `run` method in the common source module to respect the `alwaysRunInSingleProcessMode` method when deciding whether to call `processSingle` or `processCollection` ([link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-8463d90b1fce2b0a62c9996d446a3b269e680bc0598a76accdadd3ea59d268d2L201-R211))
*  Add new source modules for Snowflake that emit events when a new database, schema, or table is created, extending the common module and overriding some methods (`components/snowflake/sources/new-database/new-database.mjs`, `components/snowflake/sources/new-schema/new-schema.mjs`, `components/snowflake/sources/new-table/new-table.mjs`) ([link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-3ddbe7bea9e4ecab5889cbe3f40d2d7337e0dee42a0b7dbdf8a734299f4a73f5R1-R47), [link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-a547f7581b5437077f6235a72c6b018f42ffed1019ca24e3951b173755a2ff87R1-R47), [link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-f738b23c10442b3b965f2377fcc88fd18d679aea39da15b3bf6c611e0196d2bbR1-R50))
*  Use the `alwaysRunInSingleProcessMode` method to return true in the new source modules, since they need to update the last execution time after each run ([link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-3ddbe7bea9e4ecab5889cbe3f40d2d7337e0dee42a0b7dbdf8a734299f4a73f5R1-R47), [link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-a547f7581b5437077f6235a72c6b018f42ffed1019ca24e3951b173755a2ff87R1-R47), [link](https://github.com/PipedreamHQ/pipedream/pull/6352/files?diff=unified&w=0#diff-f738b23c10442b3b965f2377fcc88fd18d679aea39da15b3bf6c611e0196d2bbR1-R50))

### CODE REVIEW

File: components/snowflake/sources/new-database/new-database.mjs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import common from "../common.mjs";
+import { v4 as uuidv4 } from "uuid";
+
+export default {
+  ...common,
+  type: "source",
+  key: "snowflake-new-database",
+  name: "New Database",
+  description: "Emit new event when a database is created",
```
Comment (dylburger) [inline components/snowflake/sources/new-database/new-database.mjs:9]:
```suggestion
  description: "Emit a new event when a database is created",
```

File: components/snowflake/sources/new-schema/new-schema.mjs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import common from "../common.mjs";
+import { v4 as uuidv4 } from "uuid";
+
+export default {
+  ...common,
+  type: "source",
+  key: "snowflake-new-schema",
+  name: "New Schema",
+  description: "Emit new event when a schema is created",
```
Comment (dylburger) [inline components/snowflake/sources/new-schema/new-schema.mjs:9]:
```suggestion
  description: "Emit a new event when a schema is created",
```

Comment (dylburger) [reply components/snowflake/sources/new-schema/new-schema.mjs:9]: Changed "Emit new event" to "Emit a new event" to follow Google's Developer docs style guide (use articles)

File: components/snowflake/sources/new-table/new-table.mjs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import common from "../common.mjs";
+import { v4 as uuidv4 } from "uuid";
+
+export default {
+  ...common,
+  type: "source",
+  key: "snowflake-new-table",
+  name: "New Table",
+  description: "Emit new event when a table is created",
```
Comment (dylburger) [inline components/snowflake/sources/new-table/new-table.mjs:9]:
```suggestion
  description: "Emit a new event when a table is created",
```

Comment (dylburger) [reply components/snowflake/sources/new-table/new-table.mjs:9]: Changed "Emit new event" to "Emit a new event" to follow Google's Developer docs style guide (use articles)

File: components/snowflake/sources/new-database/new-database.mjs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  version: "0.0.2",
```
Comment (GTFalcao) [inline components/snowflake/sources/new-database/new-database.mjs:10]: This is a new component, right? I believe it should be at the minimum version
```suggestion
  version: "0.0.1",
```

File: components/snowflake/package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@pipedream/snowflake",
-  "version": "0.4.4",
+  "version": "0.4.5",
```
Comment (GTFalcao) [inline components/snowflake/package.json:3]: As per [the versioning guidelines](https://pipedream.com/docs/components/guidelines/#versioning), including a new feature (component) warrants a new `minor` version
```suggestion
  "version": "0.5.0",
```

**Review by GTFalcao — COMMENTED** [review]
Hi @vellames , thanks for the contribution!

I've placed a few comments regarding version numbers. Since this doesn't affect the code itself, I'm moving this forward

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Snowflake_Part1_6352_1315.pdf](https://github.com/PipedreamHQ/pipedream/files/11438650/Snowflake_Part1_6352_1315.pdf)

**Review by GTFalcao — APPROVED** [review]

**Pull request merged**